### ISSUE TITLE
does yolo12 support  pytorch2 inductor AOT mode export ?

### ISSUE THREAD

**Issue by mullerhai** [ultralytics/ultralytics#23072]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

cannot  export aot pt2 file for yolo11 yolo12 

### Environment

python 3.13  pycharm  mac m5 32gb

### Minimal Reproducible Example

```
import torch
from ultralytics import YOLO

# 1. 加载模型
model = YOLO("yolo12n.pt")
native_model = model.model.eval()

# 2. 准备伪输入
img = torch.randn(1, 3, 640, 640)

# 3. 先进行 torch.export (这是 AOTI 的前置步骤)
# 注意：YOLO 模型可能包含动态操作，这里先尝试简单导出
exported_program = torch.export.export(native_model, args=(img,))

# 4. 使用 AOTI 编译
# 根据你看到的源码，不再需要传递 args 参数，因为已经包含在 exported_program 里的
so_path = torch._inductor.aoti_compile_and_package(
    exported_program,
    package_path="yolo12_aot.pt2"
)

print(f"成功导出 AOT 模型至: {so_path}")

```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mullerhai, thank you for your interest in Ultralytics 🚀! This is an automated response to help us triage your issue quickly—an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. You’ve already shared code—thank you! 🙏 To make it a complete MRE, please also include the exact `ultralytics` version, `torch` version, and the full traceback/log output you’re seeing, plus the exact export command/settings you tried (if any) within Ultralytics’ export pipeline.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
You can use the dynamo branch and export:

https://github.com/ultralytics/ultralytics/pull/20370

```
pip install git+https://github.com/ultralytics/ultralytics@dynamo
```

```python
import torch
from ultralytics import YOLO

# 1. 加载模型
model = YOLO("yolo12n.pt")
native_model = model.model.eval()
native_model.model[-1].export = True

# 2. 准备伪输入
img = torch.randn(1, 3, 640, 640)

# 3. 先进行 torch.export (这是 AOTI 的前置步骤)
# 注意：YOLO 模型可能包含动态操作，这里先尝试简单导出
exported_program = torch.export.export(native_model, args=(img,))

# 4. 使用 AOTI 编译
# 根据你看到的源码，不再需要传递 args 参数，因为已经包含在 exported_program 里的
so_path = torch._inductor.aoti_compile_and_package(
    exported_program,
    package_path="yolo12_aot.pt2"
)

print(f"成功导出 AOT 模型至: {so_path}")
```

**Comment by mullerhai** [comment]
> You can use the dynamo branch and export:
> 
> [#20370](https://github.com/ultralytics/ultralytics/pull/20370)
> 
> ```
> pip install git+https://github.com/ultralytics/ultralytics@dynamo
> ```
> 
> import torch
> from ultralytics import YOLO
> 
> # 1. 加载模型
> model = YOLO("yolo12n.pt")
> native_model = model.model.eval()
> native_model.model[-1].export = True
> 
> # 2. 准备伪输入
> img = torch.randn(1, 3, 640, 640)
> 
> # 3. 先进行 torch.export (这是 AOTI 的前置步骤)
> # 注意：YOLO 模型可能包含动态操作，这里先尝试简单导出
> exported_program = torch.export.export(native_model, args=(img,))
> 
> # 4. 使用 AOTI 编译
> # 根据你看到的源码，不再需要传递 args 参数，因为已经包含在 exported_program 里的
> so_path = torch._inductor.aoti_compile_and_package(
>     exported_program,
>     package_path="yolo12_aot.pt2"
> )
> 
> print(f"成功导出 AOT 模型至: {so_path}")

ultralytics@dynamo is. new git branch ?  could you add this module  as new version  release  it ?  
I  cannot execute  pip install git+https://github.com/ultralytics/ultralytics@dynamo. , it is failed

**Comment by Y-T-G** [comment]
What's the error? The PR won't be merged anytime soon

**Comment by mullerhai** [comment]
> What's the error? The PR won't be merged anytime soon

sorry ，just  I wait for 30 minutes ，It is could install. sucessfully , and it could work . very thanks

**Comment by mullerhai** [comment]
HI @Y-T-G  Don't rush to close it, I have a groundbreaking breakthrough share with you here that greatly expands the scope of YOLO12. From now on, for Java users, you can fully recommend they use this method, which is much more convenient than ONNX or other methods. You can now use javacpp-pytorch to use the newly exported pt2 model file very smoothly, although I treated it as a compressed package, unpacked it, and got the internal .so files. I'll paste the code for you to see. Also, I wrote a technical blog that explains how to export YOLO11n and YOLO11x to ONNX and TorchScript files and use javacpp-pytorch to load the exported model for inference.

blog
[p.weixin.qq.com/s/hBsBZ759zhKrEd-4MZZ2yQ](https://mp.weixin.qq.com/s/hBsBZ759zhKrEd-4MZZ2yQ?token=2018697551&lang=en_US)

```
package torch.geometric.demo;

import org.bytedeco.javacpp.*;
import org.bytedeco.pytorch.*;
import static org.bytedeco.pytorch.global.torch.*;

import java.util.Arrays;

public class YOLO12AOTInference {
    public static void main(String[] args) {
        // 1. 指定 .so 模型路径 (确保路径正确)
        String modelPath ="yolo12_aot/data/aotinductor/model/cth7gpzhojzjn3pwhhdttlp7u5snht65f362t6lkxobt45qtdgjh.wrapper.so";

        // 2. 环境准备：强制单线程以避免 macOS 上的 OpenMP 冲突 (SIGSEGV)
        System.setProperty("org.bytedeco.openblas.load", "nolapack");

        // 3. 使用 AOTIModelContainerRunnerCpu 加载模型
        // 参数: 路径, 实例数=1, 强制单线程=true
        try (AOTIModelContainerRunnerCpu runner = new AOTIModelContainerRunnerCpu(modelPath, 1, true)) {

            System.out.println("✅ AOT 模型加载成功!");

            // 4. 构造输入 Tensor (Shape: 1, 3, 640, 640)
            // 注意：dtype 必须匹配导出时的 Float32
            long[] inputShape = {1, 3, 640, 640};
            Tensor inputTensor = randn(inputShape, new TensorOptions().dtype(new ScalarTypeOptional(ScalarType.Float))).contiguous();

            // 5. 封装输入为 TensorVector (AOTI 接收 Vector 容器)
            TensorVector inputs = new TensorVector(inputTensor);

            System.out.println("🚀 开始推理...");
            long startTime = System.nanoTime();

            // 6. 执行推理
            // runner.run 会返回一个 TensorVector，包含模型所有的输出节点
            TensorVector outputs = runner.run(inputs);

            long endTime = System.nanoTime();
            System.out.printf("⏱️ 推理耗时: %.2f ms%n", (endTime - startTime) / 1e6);

            // 7. 处理输出结果
            System.out.println("📦 检测到输出层数量: " + outputs.size());

            for (int i = 0; i < outputs.size(); i++) {
                Tensor out = outputs.get(i);
                long[] outShape = out.sizes().vec().get();
                System.out.println("层 " + i + " 形状: " + Arrays.toString(outShape));

                // 如果你想读取具体数值 (例如前 10 个)
                FloatPointer data = out.data_ptr_float();
                System.out.print("前 10 个原始数值: ");
                for (int j = 0; j < Math.min(10, out.numel()); j++) {
                    System.out.print(data.get(j) + " ");
                }
                System.out.println("\n---");
            }

        } catch (Exception e) {
            System.err.println("❌ AOT 推理失败:");
            e.printStackTrace();
        }
    }
}
```

the. log
```
✅ AOT 模型加载成功!
🚀 开始推理...
⏱️ 推理耗时: 140.20 ms
📦 检测到输出层数量: 1
层 0 形状: [1, 84, 8400]
前 10 个原始数值: 11.2385435 13.068897 19.624123 23.152954 35.069546 49.765987 58.049564 66.109375 67.50868 74.24699 
---

```

**Comment by Y-T-G** [comment]
@onuralpszr May be interesting

**Comment by Y-T-G** [comment]
@mullerhai I think you can just use ExecuTorch export which is intended for what you want

https://docs.ultralytics.com/integrations/executorch/

**Comment by mullerhai** [comment]
No,executorch export model only coulduse on Android for java,can't use for Ubuntu Mac windows> @mullerhai I think you can just use ExecuTorch export which is intended for what you want
> 
> https://docs.ultralytics.com/integrations/executorch/

**Comment by glenn-jocher** [comment]
ExecuTorch isn’t Android-only: the exported `.pte` is consumed by the ExecuTorch runtime, which can be built for desktop (Linux/macOS/Windows) as well as mobile/embedded; what’s missing today is an official Java API from our side, so for Java on Ubuntu/macOS/Windows you’d need to wrap the ExecuTorch C++ runtime via JNI (or keep using your `pt2`/AOTInductor + JavaCPP flow), and for export details see the Ultralytics `export_executorch` implementation in the [Exporter reference](https://docs.ultralytics.com/reference/engine/exporter/) and platform/runtime notes in the [PyTorch ExecuTorch documentation](https://docs.pytorch.org/executorch/).

**Comment by mullerhai** [comment]
> ExecuTorch isn’t Android-only: the exported `.pte` is consumed by the ExecuTorch runtime, which can be built for desktop (Linux/macOS/Windows) as well as mobile/embedded; what’s missing today is an official Java API from our side, so for Java on Ubuntu/macOS/Windows you’d need to wrap the ExecuTorch C++ runtime via JNI (or keep using your `pt2`/AOTInductor + JavaCPP flow), and for export details see the Ultralytics `export_executorch` implementation in the [Exporter reference](https://docs.ultralytics.com/reference/engine/exporter/) and platform/runtime notes in the [PyTorch ExecuTorch documentation](https://docs.pytorch.org/executorch/).

really? or you have update and release new version jar?
.expect  executorch jar dependence remove the Android system jars ,if you think  pure java project could use ,please show me one executorch with pure java code example could running ON Mac or Linux like my code,thanks

**Comment by mullerhai** [comment]
@glenn-jocher @Y-T-G   @lucylq @psiddh @ kirklandsign


<img width="467" height="963" alt="Image" src="https://github.com/user-attachments/assets/62d4c519-87e3-40bf-9c0b-b974cb890ef8" />

<img width="1282" height="773" alt="Image" src="https://github.com/user-attachments/assets/cd450bfb-6dc9-4c69-8fff-478931b24dc9" />

please tell me where to find。executorch for。linux。mac。windows with。java？

https://github.com/pytorch/executorch/issues/15688


> ExecuTorch isn’t Android-only: the exported `.pte` is consumed by the ExecuTorch runtime, which can be built for desktop (Linux/macOS/Windows) as well as mobile/embedded; what’s missing today is an official Java API from our side, so for Java on Ubuntu/macOS/Windows you’d need to wrap the ExecuTorch C++ runtime via JNI (or keep using your `pt2`/AOTInductor + JavaCPP flow), and for export details see the Ultralytics `export_executorch` implementation in the [Exporter reference](https://docs.ultralytics.com/reference/engine/exporter/) and platform/runtime notes in the [PyTorch ExecuTorch documentation](https://docs.pytorch.org/executorch/).

**Comment by glenn-jocher** [comment]
ExecuTorch doesn’t currently publish a “pure JVM” (Linux/macOS/Windows) Java `jar`—the only official Java/Kotlin bindings are shipped as an Android AAR (with JNI + native libs), as described in the upstream docs on [using ExecuTorch on Android](https://docs.pytorch.org/executorch/stable/using-executorch-android.html); for desktop Java you’d need to build the ExecuTorch C++ runtime and write your own JNI wrapper (similar to what the Android AAR provides), following the upstream [ExecuTorch setup/build guide](https://docs.pytorch.org/executorch/0.4/getting-started-setup.html), or stick with your current `pt2`/AOTInductor + JavaCPP approach.

**Comment by mullerhai** [comment]
> ExecuTorch doesn’t currently publish a “pure JVM” (Linux/macOS/Windows) Java `jar`—the only official Java/Kotlin bindings are shipped as an Android AAR (with JNI + native libs), as described in the upstream docs on [using ExecuTorch on Android](https://docs.pytorch.org/executorch/stable/using-executorch-android.html); for desktop Java you’d need to build the ExecuTorch C++ runtime and write your own JNI wrapper (similar to what the Android AAR provides), following the upstream [ExecuTorch setup/build guide](https://docs.pytorch.org/executorch/0.4/getting-started-setup.html), or stick with your current `pt2`/AOTInductor + JavaCPP approach.

If generating a Java package with executorch is this simple, why doesn't the official executorch team do it themselves? Why do they have to leave it to users to figure out? It's an excellent tool, yet they make users build it themselves. I find this extremely unreasonable.

**Comment by onuralpszr** [comment]
> No,executorch export model only coulduse on Android for java,can't use for Ubuntu Mac windows> [@mullerhai](https://github.com/mullerhai) I think you can just use ExecuTorch export which is intended for what you want
> 
> > https://docs.ultralytics.com/integrations/executorch/

Yes I understand you wanna go for full java desktop/server solution without rewriting and that's as you opened it is executorch side issue, I can try to best help but as you said their code mostly embedded into android mobile platform optimized solution and in order to make executorch and java pure happy it needs to be changed @Y-T-G  this one I leave to executorch team and issue already opened. Once they have something we can write example to run in desktop to show case and I can try to show case demo/example as well.

**Comment by mullerhai** [comment]
> > No,executorch export model only coulduse on Android for java,can't use for Ubuntu Mac windows> [@mullerhai](https://github.com/mullerhai) I think you can just use ExecuTorch export which is intended for what you want
> > > https://docs.ultralytics.com/integrations/executorch/
> 
> Yes I understand you wanna go for full java desktop/server solution without rewriting and that's as you opened it is executorch side issue, I can try to best help but as you said their code mostly embedded into android mobile platform optimized solution and in order to make executorch and java pure happy it needs to be changed [@Y-T-G](https://github.com/Y-T-G) this one I leave to executorch team and issue already opened. Once they have something we can write example to run in desktop to show case and I can try to show case demo/example as well.

hi @onuralpszr  @glenn-jocher  @Y-T-G 
Thank you so much for your help. This is the first time I've ever felt that someone was genuinely listening to me. I was extremely frustrated and angry when communicating with the executorch team. I don't understand why they would pretend not to understand when they clearly do, just to brush me off. A great project just wants to be used on the Android platform. It's not a technical limitation; they simply don't want to implement it. I am furious about this. I have been actively promoting PyTorch on Java, as well as Graph Neural Networks on Java, Reinforcement Learning on Java, and LLM inference on Java. I have no idea what the executorch team is so busy with, but they are showing clear favoritism. Is it a staffing issue, or is there no KPI for this work? No one seems to have the motivation to do it. This is by far the worst team I have had to deal with. If they are being paid by Meta, then the entire team is failing to deliver any real value.

**Comment by glenn-jocher** [comment]
I hear your frustration, but we can’t comment on (or escalate) upstream team priorities—right now there simply isn’t an official “pure JVM” ExecuTorch desktop `jar` we can point you to, so we can’t provide a working pure-Java macOS/Linux example from the Ultralytics side; the most practical path today is to keep using your `pt2`/AOTInductor + JavaCPP workflow and track progress in the upstream thread ([pytorch/executorch#15688](https://github.com/pytorch/executorch/issues/15688)). If you hit an Ultralytics-specific `.pte` export problem, please open a new Ultralytics issue with the full traceback plus:

```bash
yolo checks
```

and we’ll take a look; export usage details are in the [Ultralytics ExecuTorch integration guide](https://docs.ultralytics.com/integrations/executorch/).
